File: Bukkit/src/main/java/me/mrkirby153/kcutils/flags/FlagModule.java
package me.mrkirby153.kcutils.flags;

import me.mrkirby153.kcutils.C;
import me.mrkirby153.kcutils.Module;
import me.mrkirby153.kcutils.command.CommandManager;
import net.md_5.bungee.api.ChatColor;
import net.md_5.bungee.api.chat.TextComponent;
import org.bukkit.Sound;
import org.bukkit.SoundCategory;
import org.bukkit.World;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.Arrays;
import java.util.HashMap;

public class FlagModule extends Module<JavaPlugin> {

    private HashMap<String, FlagSettings> worldSettings = new HashMap<>();

    public FlagModule(JavaPlugin plugin) {
        super("WorldFlags", "1.0", plugin);
    }

    public boolean get(World world, WorldFlags flag) {
        if (worldSettings.containsKey(world.getName())) {
            return worldSettings.get(world.getName()).isSet(flag);
        } else {
            return flag.defaultValue();
        }
    }

    public FlagSettings getFlags(World toActOn) {
        return worldSettings.get(toActOn.getName());
    }

    public FlagSettings initialize(World world) {
        FlagSettings settings = this.worldSettings.get(world.getName());
        if (settings == null)
            settings = new FlagSettings(world.getName());
        FlagSettings finalSettings = settings;
        Arrays.stream(WorldFlags.values()).forEach(flag -> finalSettings.setFlag(flag, flag.defaultValue()));
        this.worldSettings.put(world.getName(), settings);
        return settings;
    }

    public void set(World world, WorldFlags flag, boolean state, boolean announce) {
        if (!worldSettings.containsKey(world.getName()))
            return;
        worldSettings.get(world.getName()).setFlag(flag, state);
        if (announce)
            world.getPlayers().forEach(player -> {
                TextComponent component = C.formattedChat(flag.getFriendlyName() + "> ", C.TAG_COLOR, C.Style.BOLD);

                TextComponent s = C.formattedChat(state ? "true" : "false", state ? ChatColor.GREEN : ChatColor.RED, C.Style.BOLD);
                component.addExtra(s);
                player.spigot().sendMessage(component);
                player.playSound(player.getLocation(), Sound.BLOCK_NOTE_PLING, SoundCategory.MASTER, 1F, 1F);
            });
    }

    public void set(World world, WorldFlags flag, boolean state) {
        set(world, flag, state, true);
    }

    public boolean shouldCancel(World world, WorldFlags flag) {
        return !get(world, flag);
    }

    @Override
    protected void init() {
        registerListener(new FlagListener(this));
        if (CommandManager.instance() != null)
            CommandManager.instance().registerCommand(new FlagCommand(getPlugin(), this));
        else
            log("[WARN] Command manager not initialized! "+FlagCommand.class+" needs to be registered manually!");
    }
}

File: Bukkit/src/main/kotlin/me/mrkirby153/kcutils/flags/FlagModule.kt
package me.mrkirby153.kcutils.flags

import me.mrkirby153.kcutils.Chat
import me.mrkirby153.kcutils.Module
import me.mrkirby153.kcutils.command.CommandManager
import net.md_5.bungee.api.ChatColor
import org.bukkit.Sound
import org.bukkit.SoundCategory
import org.bukkit.World
import org.bukkit.plugin.java.JavaPlugin
import java.util.*

/**
 * A module for handling various flags on the world
 *
 * @param plugin    The owning plugin
 */
class FlagModule(plugin: JavaPlugin) : Module<JavaPlugin>("WorldFlags", plugin) {

    private val worldSettings = HashMap<String, FlagSettings>()

    /**
     * Gets a flag's status for the world
     *
     * @param world The world
     * @param flag  The flag
     *
     * @return True if the flag is set
     */
    operator fun get(world: World, flag: WorldFlags): Boolean {
        return worldSettings[world.name]?.isSet(flag) ?: flag.defaultValue()
    }

    /**
     * Gets all set flags on the world
     *
     * @param toActOn The world
     *
     * @return The flag settings
     */
    fun getFlags(toActOn: World): FlagSettings {
        return worldSettings[toActOn.name] ?: FlagSettings(toActOn.name)
    }

    /**
     * Initialize the world by setting all flags to their default values
     *
     * @param world The world to initialize
     *
     * @return The world's flag settings
     */
    fun initialize(world: World): FlagSettings {
        var settings: FlagSettings? = this.worldSettings[world.name]
        if (settings == null)
            settings = FlagSettings(world.name)
        val finalSettings = settings
        Arrays.stream(WorldFlags.values()).forEach { flag -> finalSettings.setFlag(flag, flag.defaultValue()) }
        this.worldSettings.put(world.name, settings)
        return settings
    }

    /**
     * Checks if the world has been initialized
     *
     * @param world The World to check
     *
     * @return True if the world has had its flags initialized
     */
    fun initialized(world: World): Boolean {
        return this.worldSettings.containsKey(world.name)
    }

    /**
     * Sets a flag in the world
     *
     * @param world     The world to set
     * @param flag      The flag to set
     * @param state     The state of the flag (enabled/disabled)
     * @param announce  If the flag change is to be announced
     */
    @JvmOverloads
    fun set(world: World, flag: WorldFlags, state: Boolean, announce: Boolean = true) {
        if (!worldSettings.containsKey(world.name))
            return
        worldSettings[world.name]?.setFlag(flag, state)
        if (announce)
            world.players.forEach { player ->
                val component = Chat.formattedChat(flag.friendlyName + "> ", Chat.TAG_COLOR, Chat.Style.BOLD)

                val s = Chat.formattedChat(if (state) "true" else "false", if (state) ChatColor.GREEN else ChatColor.RED, Chat.Style.BOLD)
                component.addExtra(s)
                player.spigot().sendMessage(component)
                player.playSound(player.location, Sound.BLOCK_NOTE_PLING, SoundCategory.MASTER, 1f, 1f)
            }
    }

    /**
     * Checks if the flag should be canceled
     *
     * @param world The world to check
     * @param flag  The flag to check
     *
     * @return True if the event should be canceled
     */
    fun shouldCancel(world: World, flag: WorldFlags): Boolean {
        return !get(world, flag)
    }

    override fun init() {
        registerListener(FlagListener(this))
        if (CommandManager.instance() != null)
            CommandManager.instance()?.registerCommand(FlagCommand(plugin, this))
        else
            log("[WARN] Command manager not initialized! " + FlagCommand::class.java + " needs to be registered manually!")
    }
}
